### PR TITLE
python3Packages.*Hook: support `__structuredAttrs = true` (the easier ones)

### PR DESCRIPTION
**Pull request by ShamrockLee**
This are the easier, less controversial part of the effort to make `buildPythonPackage` and `buildPythonApplication` support both `__structuredAttrs = true` and `structuredAttrs = false`.

Each modified Bash script is linted with ShellCheck.

It also fixes some leftovers of #339117 among Python-related hooks.

Split from #347194

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by wolfgangwalther — CHANGES_REQUESTED** [review]
Just read through, so far. Now we need packages that actually make use of those flags and test them with structuredAttrs and without:

- `pipBuildFlags`... is not used at all, so no useful package to test.
- `pipInstallFlags`.. used 3 times, but only with a single flag. Not useful to test.
- `pypaBuildFlags` could be tested with `pillow-jpls`. This should fail with structuredAttrs on master/staging, if I'm not mistaken, but pass with structuredAttrs in this PR.
- `pythonImportsCheck` should be tested carefully with one of the many packages that has multiple items set. This would likely fail silently with structuredAttrs on master/staging, only testing the first list item, but would test all of them with this branch.
- `pythonNamespaces` also needs to be checked manually, to confirm the cleanup was in fact broken when multiple namespaces were given with structuredAttrs, but is fixed now.
- `pythonRelaxDeps` and `pythonRemoveDeps` should equally be tested manually - the build with structuredAttrs would probably not outright fail, yet, but the metadata file be subtly broken. Should be fixed with this branch.

File: pkgs/development/interpreters/python/hooks/setuptools-rust-hook.sh
Comment (ShamrockLee) [reply pkgs/development/interpreters/python/hooks/setuptools-rust-hook.sh:20]: I just checked `runHook` in `setup.sh`, and *Hooks are meant to be Bash arrays is never passed as attributes.

You are right. I should not change them.

File: pkgs/development/interpreters/python/hooks/python-relax-deps-hook.sh
Comment (ShamrockLee) [reply pkgs/development/interpreters/python/hooks/python-relax-deps-hook.sh:44]: ShellCheck should not print warnings in the source file except for parsing errors. As for the out-of-lint `setup.sh`, I created #351849 to correct it.

> Let's just remove that for now, until we figure out a full solution with CI, which makes sure to have generic/setup.sh be proper as well?

That makes sense.

File: pkgs/development/python-modules/meson-python/add-build-flags.sh
Comment (ShamrockLee) [reply pkgs/development/python-modules/meson-python/add-build-flags.sh:6]: Do you mean I should append `pypaBuildFlags` to `pipBuildFlags`?

Comment (ShamrockLee) [reply pkgs/development/python-modules/meson-python/add-build-flags.sh:6]: Ahh! That line slipped off my eyes. Thank you for reminding me.

Comment (ShamrockLee) [reply pkgs/development/python-modules/meson-python/add-build-flags.sh]: Addressed.

File: pkgs/development/interpreters/python/hooks/setuptools-rust-hook.sh
Comment (ShamrockLee) [reply pkgs/development/interpreters/python/hooks/setuptools-rust-hook.sh]: Applied.

**Comment by ShamrockLee** [comment]
> `pypaBuildFlags` could be tested with `pillow-jpls`. This should fail with structuredAttrs on master/staging, if I'm not mistaken, but pass with structuredAttrs in this PR.

`pillow-jpls` has a strange `pypaBuildFlags` specification, which seems to expect Bash expansion before execution. However, `pypaBuildHook` does not evaluate the constructed command before execution like `pytestCheckHook` does. I wonder if `pillow-jpls` would ever build as expected by the maintainer.

Cc: @teh (`python3Packages.pillow-jpls` maintainer)

**Comment by ShamrockLee** [comment]
Update: `python3Packages.pillow-jpls` works for both `__structuredAttrs = true` and `structuredAttrs = false`, but the build logs of both cases contain the following message:

```
*** Configuring CMake...
CMake Warning:
  Ignoring extra path from command line:

   "'--preset=sysdeps'"

```

This line is also in the build log of `python3Packages.pillow-jpls` on the `master` branch.

**Comment by ShamrockLee** [comment]
> I tested it and now I understand what you mean: Because the python code accesses this variable as an **environment** variable it works without structuredAttrs. But with structuredAttrs this is not an exported environment variable anymore, but just a bash variable - thus python can't read it, and the imports check fails indeed.

Exactly.

**Review by wolfgangwalther — COMMENTED** [review]
I tested all commits, they do what they claim to do - fix those hooks with structuredAttrs turned on.

While the sum of changes is good, the split across commits is still not 100% clean. For some hooks shellcheck and structuredAttrs changes claim to be split between two commits, but the structuredAttrs commit contains part of the shellcheck changes.

The first commit should better be `treewide:` and fix the remaining cases of that the `*Phases+=` pattern as well.

The namespace hook should be a single shellcheck-only commit.

Other than that - it works.

**Review by wolfgangwalther — APPROVED** [review]
This needs a rebase for the base branch to evaluate.

LGTM otherwise!

Thanks for pushing structuredAttrs :)

File: pkgs/development/interpreters/python/hooks/pip-install-hook.sh
Comment (emilazy) [inline pkgs/development/interpreters/python/hooks/pip-install-hook.sh:10]: Not blocking, but: would be nice to do a trick like that `/dev/null` thing to get declarations of all these essential variables.

File: pkgs/development/interpreters/python/hooks/pypa-build-hook-test.nix
Comment (emilazy) [inline pkgs/development/interpreters/python/hooks/pypa-build-hook-test.nix:15]: Is this needed, or just general clean‐up of the example?

File: pkgs/development/interpreters/python/hooks/python-imports-check-hook.sh
Comment (emilazy) [inline pkgs/development/interpreters/python/hooks/python-imports-check-hook.sh:15]: Unnecessary change, right? (But perfectly fine: I just want to check I’m not missing a nuance.)

File: pkgs/development/interpreters/python/hooks/python-namespaces-hook.sh
Comment (emilazy) [inline pkgs/development/interpreters/python/hooks/python-namespaces-hook.sh:17]: I know that Python namespace names are restricted enough that this is fine, but I’m not in love with moving from things like `read` to these more interpolation‐y solutions. Was ShellCheck unhappy with the previous state?

File: pkgs/development/interpreters/python/hooks/python-namespaces-hook.sh
Comment (emilazy) [inline pkgs/development/interpreters/python/hooks/python-namespaces-hook.sh:40]: Is there a reason to move to `[[ … ]]` here but not in the other hooks?

File: pkgs/development/interpreters/python/hooks/python-remove-tests-dir-hook.sh
Comment (emilazy) [reply pkgs/development/interpreters/python/hooks/python-remove-tests-dir-hook.sh:11]: Ironically non‐shell‐safe `$out` paths are way more hopelessly broken than doing it inside a package could ever be. But yeah, would be nice, but not blocking.

**Review by emilazy — APPROVED** [review]
This looks good and indeed easy to me! Some questions to make sure I understand everything fully.

File: pkgs/development/interpreters/python/hooks/python-namespaces-hook.sh
Comment (ShamrockLee) [reply pkgs/development/interpreters/python/hooks/python-namespaces-hook.sh:12]: I squashed those changes into the `pythonNamespaceHook: lint with ShellCheck` commit.

File: pkgs/development/interpreters/python/hooks/pypa-build-hook-test.nix
Comment (ShamrockLee) [reply pkgs/development/interpreters/python/hooks/pypa-build-hook-test.nix:15]: This change makes it closer to how an actual project would be structured. The resulting package after this change could be Python-imported like `import my_project`.

Such change is not strictly necessary to build the test anyway.

File: pkgs/development/interpreters/python/hooks/python-namespaces-hook.sh
Comment (ShamrockLee) [reply pkgs/development/interpreters/python/hooks/python-namespaces-hook.sh:17]: Oops! I cannot recall why I made this change. The `read` one looks better. I'll change it back.

File: pkgs/development/interpreters/python/hooks/python-namespaces-hook.sh
Comment (ShamrockLee) [reply pkgs/development/interpreters/python/hooks/python-namespaces-hook.sh:40]: I just forgot to move to `[[  ... ]]` when linting `pythonImportsCheckHook`.

File: pkgs/development/interpreters/python/hooks/python-output-dist-hook.sh
Comment (ShamrockLee) [reply pkgs/development/interpreters/python/hooks/python-output-dist-hook.sh:27]: > I'll clean them in another PR.

I opened #352709 to address them.

File: pkgs/development/interpreters/python/hooks/python-imports-check-hook.sh
Comment (emilazy) [reply pkgs/development/interpreters/python/hooks/python-imports-check-hook.sh:15]: Interesting! I thought that `foo=$bar` didn’t ever need any quoting – or is it just a stylistic lint?

**Comment by philiptaron** [comment]
Should we merge this in the next couple days?

**Comment by emilazy** [comment]
I was waiting for @ShamrockLee to rebase per responses to review feedback before merging.

File: pkgs/development/interpreters/python/hooks/python-imports-check-hook.sh
Comment (ShamrockLee) [reply pkgs/development/interpreters/python/hooks/python-imports-check-hook.sh:15]: It is unnecessary.

File: pkgs/development/interpreters/python/hooks/python-namespaces-hook.sh
Comment (ShamrockLee) [reply pkgs/development/interpreters/python/hooks/python-namespaces-hook.sh:39]: > I just forgot to move to `[[ ... ]]` when linting `pythonImportsCheckHook`.

Applied.

**Comment by ShamrockLee** [comment]
Rebased.

**Comment by emilazy** [comment]
LGTM, thanks!

**Pull request merged**
